File: analyzers/tests/SonarAnalyzer.Test/TestCases/GenericInheritanceShouldNotBeRecursive.vb
Imports System

Namespace Tests.ClassRecursion

    ' some base classes
    Class CA(Of T)
    End Class

    Class CB(Of T)
    End Class

    Class CC(Of T1, T2)
    End Class

    Class C0(Of T)
        Inherits CA(Of C0(Of T))
    End Class

    Class C1(Of T) ' Compliant
        Inherits CA(Of CB(Of C1 (Of T)))
    End Class

    Class C2(Of T) ' Noncompliant {{Refactor this Class so that the generic inheritance chain is not recursive.}}
'         ^^
        Inherits CA(Of C2(Of CB(Of T)))
    End Class

    Class C3(Of T) ' Noncompliant
        Inherits CA(Of C3(Of C3(Of T)))
    End Class

    Class C4(Of T) ' Noncompliant
        Inherits CA(Of C4(Of CA(Of T)))
    End Class

    Class C5(Of T) ' Noncompliant
        Inherits CC(Of C5(Of CA(Of T)), CB(Of T))
    End Class

    Class C6(Of T) ' Compliant
        Inherits CC(Of C5(Of CA(Of T)), CB(Of T))
    End Class

    Class C7(Of T) ' Noncompliant
        Inherits CA(Of CA(Of CA(Of CA(Of CA(Of CA(Of CA(Of CA(Of CA(Of CA(Of C7(Of CB(Of T))))))))))))
    End Class

    Class C8(Of T) ' Noncompliant
        Implements IComparable(Of C8(Of IEquatable(Of T)))
        Public Function CompareTo(other As C8(Of IEquatable(Of T))) As Integer Implements IComparable(Of C8(Of IEquatable(Of T))).CompareTo
            Throw New NotImplementedException()
        End Function
    End Class
End Namespace

Namespace Tests.InterfaceRecursion

    ' some base classes
    Interface IA(Of T)
    End Interface

    Interface IB(Of T)
    End Interface

    Interface I0(Of T) ' Compliant
        Inherits IA(Of I0(Of T))
    End Interface

    Interface I1(Of T) ' Compliant
        Inherits IA(Of IB(Of I1 (Of T)))
    End Interface

    Interface I2(Of T) ' Compliant
        Inherits I0(Of T), IA(Of IB(Of I2(Of T)))
    End Interface

    Interface I3(Of T) ' Noncompliant
        Inherits IA(Of I3(Of IB(Of T)))
    End Interface

    Interface I4(Of T) ' Noncompliant  {{Refactor this Interface so that the generic inheritance chain is not recursive.}}
        Inherits IA(Of I4(Of I4(Of T)))
    End Interface

    Interface I5(Of T) ' Noncompliant
        Inherits IA(Of I5(Of IA(Of T)))
    End Interface

    Interface I6(Of T) ' Noncompliant
        Inherits IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of IA(Of I6(Of IB(Of T))))))))))))))
    End Interface
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/GenericInheritanceShouldNotBeRecursive.cs
using System;

namespace Tests.ClassRecursion
{
    // some base classes
    class CA<T> { }
    class CB<T> { }
    class CC<T1, T2> { }

    class C0<T> : CA<C0<T>> { } // Compliant

    class C1<T> : CA<CB<C1<T>>> { } // Compliant

    class C2<T> : CA<C2<CB<T>>> { } // Noncompliant {{Refactor this class so that the generic inheritance chain is not recursive.}}
//        ^^

    class C3<T> : CA<C3<C3<T>>> { } // Noncompliant

    class C4<T> : CA<C4<CA<T>>> { } // Noncompliant

    class C5<T> : CC<C5<CA<T>>, CB<T>> { } // Noncompliant

    class C6<T> : CC<C5<CA<T>>, CB<T>> { } // Compliant

    class C7<T> : CA<CA<CA<CA<CA<CA<CA<CA<CA<CA<CA<CA<C7<CB<T>>>>>>>>>>>>>> { } // Noncompliant

    // Error@+1 [CS0535]
    class C8<T> : IComparable<C8<IEquatable<T>>> { } // Noncompliant
}

namespace Tests.InterfaceRecursion
{
    // some base classes
    interface IA<T> { }
    interface IB<T> { }

    interface I0<T> : IA<I0<T>> { } // Compliant

    interface I1<T> : IA<IB<I1<T>>> { } // Compliant

    interface I2<T> : I0<int>, IA<IB<I2<T>>> { } // Compliant

    interface I3<T> : IA<I3<IB<T>>> { } // Noncompliant  {{Refactor this interface so that the generic inheritance chain is not recursive.}}

    interface I4<T> : IA<I4<I4<T>>> { } // Noncompliant

    interface I5<T> : IA<I5<IA<T>>> { } // Noncompliant

    interface I6<T> : IA<IA<IA<IA<IA<IA<IA<IA<IA<IA<IA<IA<I6<IB<T>>>>>>>>>>>>>> { } // Noncompliant
}
